index.js

```diff
@@ -502,6 +506,448 @@
+
+  //# ap :: Apply f => f (a -> b) -> f a -> f b
+  //.
+  //. Curried version of [`Z.ap`][].
+  //.
+  //. ```javascript
+  //. > S.ap([Math.sqrt, x => x * x], [1, 4, 9, 16, 25])
+  //. [1, 2, 3, 4, 5, 1, 16, 81, 256, 625]
+  //.
+  //. > S.ap(s => n => s.slice(0, n), s => Math.ceil(s.length / 2))('Haskell')
+  //. 'Hask'
+  //.
+  //. > S.ap(S.Just(Math.sqrt), S.Just(64))
+  //. Just(8)
+  //. ```
+  S.ap =
+  def('ap',
+      {f: [Z.Apply]},
+      [f(Fn(a, b)), f(a), f(b)],
+      Z.ap);
+
+  //# apFirst :: Apply f => f a -> f b -> f a
+  //.
+  //. Curried version of [`Z.apFirst`][]. Combines two effectful actions,
+  //. keeping only the result of the first. Equivalent to Haskell's `(<*)`
+  //. function.
+  //.
+  //. See also [`apSecond`](#apSecond).
+  //.
+  //. ```javascript
+  //. > S.apFirst([1, 2], [3, 4])
+  //. [1, 1, 2, 2]
+  //.
+  //. > S.apFirst(S.Just(1), S.Just(2))
+  //. Just(1)
+  //. ```
+  S.apFirst = def('apFirst', {f: [Z.Apply]}, [f(a), f(b), f(a)], Z.apFirst);
+
+  //# apSecond :: Apply f => f a -> f b -> f b
+  //.
+  //. Curried version of [`Z.apSecond`][]. Combines two effectful actions,
+  //. keeping only the result of the second. Equivalent to Haskell's `(*>)`
+  //. function.
+  //.
+  //. See also [`apFirst`](#apFirst).
+  //.
+  //. ```javascript
+  //. > S.apSecond([1, 2], [3, 4])
+  //. [3, 4, 3, 4]
+  //.
+  //. > S.apSecond(S.Just(1), S.Just(2))
+  //. Just(2)
+  //. ```
+  S.apSecond = def('apSecond', {f: [Z.Apply]}, [f(a), f(b), f(b)], Z.apSecond);
+
+  //# of :: Applicative f => TypeRep f -> a -> f a
+  //.
+  //. Curried version of [`Z.of`][].
+  //.
+  //. ```javascript
+  //. > S.of(Array, 42)
+  //. [42]
+  //.
+  //. > S.of(Function, 42)(null)
+  //. 42
+  //.
+  //. > S.of(S.Maybe, 42)
+  //. Just(42)
+  //.
+  //. > S.of(S.Either, 42)
+  //. Right(42)
+  //. ```
+  S.of =
+  def('of',
+      {f: [Z.Applicative]},
+      [TypeRep($.TypeVariable('f')), a, f(a)],
+      Z.of);
+
+  //# chain :: Chain m => (a -> m b) -> m a -> m b
+  //.
+  //. Curried version of [`Z.chain`][].
+  //.
+  //. ```javascript
+  //. > S.chain(x => [x, x], [1, 2, 3])
+  //. [1, 1, 2, 2, 3, 3]
+  //.
+  //. > S.chain(n => s => s.slice(0, n), s => Math.ceil(s.length / 2))('slice')
+  //. 'sli'
+  //.
+  //. > S.chain(S.parseInt(10), S.Just('123'))
+  //. Just(123)
+  //.
+  //. > S.chain(S.parseInt(10), S.Just('XXX'))
+  //. Nothing
+  //. ```
+  S.chain = def('chain', {m: [Z.Chain]}, [Fn(a, m(b)), m(a), m(b)], Z.chain);
+
+  //# join :: Chain m => m (m a) -> m a
+  //.
+  //. Alias of [`Z.join`][]. Removes one level of nesting from a nested monadic
+  //. structure.
+  //.
+  //. ```javascript
+  //. > S.join([[1], [2], [3]])
+  //. [1, 2, 3]
+  //.
+  //. > S.join([[[1, 2, 3]]])
+  //. [[1, 2, 3]]
+  //.
+  //. > S.join(S.Just(S.Just(1)))
+  //. S.Just(1)
+  //. ```
+  S.join = def('join', {m: [Z.Chain]}, [m(m(a)), m(a)], Z.join);
+
+  //# chainRec :: ChainRec m => TypeRep m -> (a -> Either (m b) (m b)) -> a -> m b
```

Comment:
type from sanctuary-type-classes is:
```hs
chainRec :: ChainRec m => (TypeRep m, (a -> c, b -> c, a) -> m c, a) -> m b
```

If we want to use `Either`, then it should be:
```hs
chainRec :: (TypeRep m, a -> m (Either a b), a) -> m b
```
in recent version of [purescript-tailRec](https://github.com/purescript/purescript-tailrec/blob/master/src/Control/Monad/Rec/Class.purs#L56) `Step` is used instead of `Either`, like this:
```hs
data Step a b = Loop a | Done b
tailRecM :: forall a b. (a -> m (Step a b)) -> a -> m b
```
---

> ... would be more consistent with the type of `chain` ...

type of `unfold` is:
```hs
:: (b -> Maybe (Pair a b)) ->       b -> Array a
```
instead of
```hs
:: (b -> Maybe (Pair a b)) -> Array b -> Array a
```

the `chainRec` looks more like unfold rather than `chain`:
```hs
unfold   :: (a -> Maybe (Pair b a)) -> a -> Array b
chainRec :: (a ->     m (Step a b)) -> a -> m     b
```

also if seed value must be of type `m a` then it means that in order for `chainRec` to be useful for arbitrary some seed `a` then the structure should also have a way to to lift  the  `a` to  `m a` i.e. `of` method, which makes the Chain a Monad. by not asking for `m a` the `chainRec` could be useful for `a`(any) value rather than `m a`. and why ask for more when it is not needed. if someone has m a  one could call chain and get `a` and in there do `chainRec`

so if we use Either it type should be
```hs
chainRec :: ChainRec m => TypeRep m -> (a -> m (Either a b)) -> a -> m b
```
and if i'm correct
```js
def('chainRec',
      {m: [Z.ChainRec]},
      [TypeRep($.TypeVariable('m')), Fn(a, m($Either(a, b))), a, m(b)],
...
```